salt/states/file.py

```diff
@@ -557,7 +557,18 @@ def process_symlinks(filenames, symlinks):
             managed_directories.add(mdest)
             keep.add(mdest)
 
-    return managed_files, managed_directories, managed_symlinks, keep
+    # Sets are randomly ordered. We need to make sure symlinks are always at the
+    # end, so we need to use a list
+    new_managed_files = list(managed_files)
+    # Now let's move all the symlinks to the end
+    for symlink in managed_symlinks:
+        for file in managed_files:
+            if file[0].endswith(os.sep + symlink[0]):
```

Comment:
I think we should be using the link target not the link path. We need to check if the target of the link matches any of the files we are creating.

Reply:
Matching full paths now, that should address this issue.